lib/chef/resource/chef_acl.rb

```diff
@@ -64,6 +57,439 @@ def remove_rights(*values)
           @remove_rights << args
         end
       end
+
+      action :create do
+        if new_resource.remove_rights && new_resource.complete
+          Chef::Log.warn("'remove_rights' is redundant when 'complete' is specified: all rights not specified in a 'rights' declaration will be removed.")
+        end
+        # Verify that we're not destroying all hope of ACL recovery here
+        if new_resource.complete && (!new_resource.rights || !new_resource.rights.any? { |r| r[:permissions].include?(:all) || r[:permissions].include?(:grant) })
+          # NOTE: if superusers exist, this should turn into a warning.
+          raise "'complete' specified on chef_acl resource, but no GRANT permissions were granted.  I'm sorry Dave, I can't let you remove all access to an object with no hope of recovery."
+        end
+
+        # Find all matching paths so we can update them (resolve * and **)
+        paths = match_paths(new_resource.path)
+        if paths.size == 0 && !new_resource.path.split('/').any? { |p| p == '*' }
+          raise "Path #{path} cannot have an ACL set on it!"
+        end
+
+        # Go through the matches and update the ACLs for them
+        paths.each do |path|
+          create_acl(path)
+        end
+      end
+
+      private
+
+      action_class.class_eval do
```

Comment:
I think there are a lot of changes we can and should make after moving it over, but was just trying to get it moved in bulk before that sort of surgery. All changes here are purely cosmetic.

There is only ever one provider for a single resource. If you fork a resource to different OS's, you just create new subclass resources.